pkg/rpc/metrics.go

```diff
@@ -108,33 +112,130 @@ is reset to zero.
+	for i := length; i < localLen; i++ {
+		childLabels = append(childLabels, "")
+	}
+	return childLabels
+}
+
+func newMetrics(locality roachpb.Locality) *Metrics {
 	childLabels := []string{"remote_node_id", "remote_addr", "class"}
-	return Metrics{
+	localityLabels := []string{}
+	for _, tier := range locality.Tiers {
+		localityLabels = append(localityLabels, tier.Key)
+	}
+	m := Metrics{
+		locality:                      locality,
 		ConnectionHealthy:             aggmetric.NewGauge(metaConnectionHealthy, childLabels...),
 		ConnectionUnhealthy:           aggmetric.NewGauge(metaConnectionUnhealthy, childLabels...),
 		ConnectionInactive:            aggmetric.NewGauge(metaConnectionInactive, childLabels...),
 		ConnectionHealthyFor:          aggmetric.NewGauge(metaConnectionHealthyNanos, childLabels...),
 		ConnectionUnhealthyFor:        aggmetric.NewGauge(metaConnectionUnhealthyNanos, childLabels...),
 		ConnectionHeartbeats:          aggmetric.NewCounter(metaConnectionHeartbeats, childLabels...),
 		ConnectionFailures:            aggmetric.NewCounter(metaConnectionFailures, childLabels...),
+		ConnectionConnected:           aggmetric.NewGauge(metaConnectionConnected, localityLabels...),
+		ConnectionBytesSent:           aggmetric.NewCounter(metaNetworkBytesEgress, localityLabels...),
+		ConnectionBytesRecv:           aggmetric.NewCounter(metaNetworkBytesIngress, localityLabels...),
 		ConnectionAvgRoundTripLatency: aggmetric.NewGauge(metaConnectionAvgRoundTripLatency, childLabels...),
 	}
+	m.mu.peerMetrics = make(map[string]peerMetrics)
+	m.mu.localityMetrics = make(map[string]localityMetrics)
+	return &m
+}
+
+type ThreadSafeMovingAverage struct {
+	syncutil.Mutex
+	ma ewma.MovingAverage
+}
+
+func (t *ThreadSafeMovingAverage) Set(v float64) {
+	t.Lock()
+	defer t.Unlock()
+	t.ma.Set(v)
+}
+
+func (t *ThreadSafeMovingAverage) Add(v float64) {
+	t.Lock()
+	defer t.Unlock()
+	t.ma.Add(v)
+}
+
+func (t *ThreadSafeMovingAverage) Value() float64 {
+	t.Lock()
+	defer t.Unlock()
+	return t.ma.Value()
 }
 
 // Metrics is a metrics struct for Context metrics.
 // Field X is documented in metaX.
 type Metrics struct {
+	locality                      roachpb.Locality
 	ConnectionHealthy             *aggmetric.AggGauge
 	ConnectionUnhealthy           *aggmetric.AggGauge
 	ConnectionInactive            *aggmetric.AggGauge
 	ConnectionHealthyFor          *aggmetric.AggGauge
 	ConnectionUnhealthyFor        *aggmetric.AggGauge
 	ConnectionHeartbeats          *aggmetric.AggCounter
 	ConnectionFailures            *aggmetric.AggCounter
+	ConnectionConnected           *aggmetric.AggGauge
+	ConnectionBytesSent           *aggmetric.AggCounter
+	ConnectionBytesRecv           *aggmetric.AggCounter
 	ConnectionAvgRoundTripLatency *aggmetric.AggGauge
+	mu                            struct {
+		syncutil.Mutex
+		// peerMetrics is a map of peerKey to peerMetrics.
+		peerMetrics map[string]peerMetrics
```

Comment:
If we are no longer changing peerMetrics, do we still need this map of metics and the mutex changes to the moving average latency?

Reply:
The map of metrics helps by preventing the "restart" of metrics when we disconnect and then reconnect. This was what you had mentioned last week and it actually makes the entire flow simpler since we can keep them in a map.

I was able to remove the reference counting and release functions completely by making the metrics more precise (we are always at 0 for the metrics we care about. We don't necessarily need the mutex around the moving average but I expect that now that we have the ability to do arbitrary aggregation of the metrics it is cleaner. I pushed a new set of smaller commits, but I'm going to wait until they pass on CI and do a simple test to make sure it all looks good.